File: src/Day02.java
import java.util.List;

public class Day02 {
    static final String FILENAME = "./src/inputs/Day02.txt";

    static List<String> passwordPolicies = Utils.readLines(FILENAME);

    static class Policy {
        private final int max;
        private final int min;
        private final char letter;
        private final String password;

        Policy(String str) {
            var policyAndPass = str.split(": ");
            this.letter = policyAndPass[0].split(" ")[1].charAt(0);
            this.min = Integer.parseInt(policyAndPass[0].split(" ")[0].split("-")[0]);
            this.max = Integer.parseInt(policyAndPass[0].split(" ")[0].split("-")[1]);
            this.password = policyAndPass[1];
        }

        public int getMax() {
            return max;
        }

        public int getMin() {
            return min;
        }

        public char getLetter() {
            return letter;
        }

        public String getPassword() {
            return password;
        }
    }

    static int part01() {
        var validCount = 0;
        for (var str : passwordPolicies) {
            var policy = new Policy(str);
            var letterCount = 0;
            for (var ch : policy.getPassword().toCharArray()) {
                if (ch == policy.getLetter())
                    letterCount++;
            }
            if (letterCount >= policy.getMin() && letterCount <= policy.getMax())
                validCount++;
        }
        return validCount;
    }

    static int part02() {
        var validCount = 0;
        for (var str : passwordPolicies) {
            var policy = new Policy(str);
            if ((policy.getLetter() == policy.getPassword().charAt(policy.getMin() - 1)) ^ (policy.getLetter() == policy.getPassword().charAt(policy.getMax() - 1)))
                validCount++;
        }
        return validCount;
    }

    public static void main() {
        System.out.println("Advent of Code 2020, Day 02");
        System.out.println(part01());
        System.out.println(part02());
    }
}

File: src/day02/Day02.kt
package day02

import java.io.File

const val FILENAME = "src/day02/Input.txt"

val passwordPolicies = File(FILENAME).readLines()

class Policy(str: String) {
    val min: Int
    val max: Int
    val letter: Char
    val password: String

    init {
        val policyAndPass = str.split(": ")
        password = policyAndPass[1]
        letter = policyAndPass[0].split(" ")[1].single()
        min = policyAndPass[0].split(" ")[0].split("-")[0].toInt()
        max = policyAndPass[0].split(" ")[0].split("-")[1].toInt()
    }
}

fun part01(): Int {
    var validCount = 0
    for (str in passwordPolicies) {
        val policy = Policy(str)
        val letterCount = policy.password.count { policy.letter == it }
        if (letterCount >= policy.min && letterCount <= policy.max)
            validCount++
    }
    return validCount
}

fun part02(): Int {
    var validCount = 0
    for (str in passwordPolicies) {
        val policy = Policy(str)
        if ((policy.letter == policy.password[policy.min - 1]) xor (policy.letter == policy.password[policy.max - 1]))
            validCount++
    }
    return validCount
}

fun main() {
    println("Advent of Code 2020, Day 02")
    println(part01())
    println(part02())
}
